src/gui/core/window_builder.cpp

```diff
@@ -153,53 +155,72 @@ builder_widget_ptr create_builder_widget(const config& cfg)
- * while linking. It seems that all these classes aren't explicitly
- * instantiated but only implicitly. Also when looking at the symbols in
- * libwesnoth-game.a the repeating button is there regardless of this #if but
- * in the final binary only if the #if is enabled.
- *
- * If this code is executed, which it will cause an assertion failure.
- *
- * Its likeley that this happens because some build this as a library file
- * which is then used by the wesnoth executable. For msvc a good try to fix
- * this issue is to add __pragma(comment(linker, "/include:" #TYPE)) or
- * similar in the REGISTER_WIDGET3 macro. For gcc and similar this can only
- * be fixed by using --whole-archive flag when linking this library.
- */
-#if 1
-#define TRY(name)                                                                                                      \
-	do {                                                                                                               \
-		if(const config& c = cfg.child(#name)) {                                                                       \
-			builder_widget_ptr p = std::make_shared<implementation::builder_##name>(c);                                \
-			assert(false);                                                                                             \
-		}                                                                                                              \
-	} while(0)
-
-	TRY(stacked_widget);
-	TRY(scrollbar_panel);
-	TRY(horizontal_scrollbar);
-	TRY(repeating_button);
-	TRY(vertical_scrollbar);
-	TRY(label);
-	TRY(image);
-	TRY(toggle_button);
-	TRY(slider);
-	TRY(scroll_label);
-	TRY(matrix);
-	TRY(minimap);
-	TRY(button);
-	TRY(menu_button);
-	TRY(drawing);
-	TRY(password_box);
-	TRY(unit_preview_pane);
-	TRY(size_lock);
-#undef TRY
-#endif
+	for(const auto& item : builder_widget_lookup()) {
+		if(item.first == "window" || item.first == "tooltip") {
+			continue;
+		}
+		if(const config& c = cfg.child(item.first)) {
+			return item.second(c);
+		}
+	}
+
+	// Force all GUI2 Widgets to be linked into Wesnoth
+	//
+	// We use a library. The linker will only include a module from the library if there is
+	// actually a reference to it. While many widgets are referenced elsewhere, it is **NOT**
+	// the case for some. The following block of code never executes, but the optimize must
+	// **NOT** see that! So we test that the cfg reference is a nullptr (references can never
+	// be a nullptr). The FORCE_LINKAGE macro creates a reference to the builder for each GUI2
+	// Widget to force the linker to include the module. So, this is all "dead code" but the
+	// compiler should never see that! Just to prove it, we'll assert false is true to crash
+	// if we ever try to execute this dead code.
+	//
+	// If you add a new GUI2 Widget, be sure to include it here, and add it's header so it's
+	// defined.
+
+	if (!cfg) { // Always false. The following never runs.
```

Comment:
Could this be `if(false)`?

Reply:
No, because the optimizer would see that it's always false and remove the code from the binary.